## *Note from the author:* 

Hey guys! This is my second Kaggle project where I decided to leverage my knowledge of EDA, ETL, Data Visualization (Tableau), SQL (MySQL), Python (NumPy, Pandas, Jupyter Notebook), and Markdown **to find out more about our X-Education software Persona qualities.** 

Additionally, I used my domain knowledge, as I worked as a **[Digital Marketing Specialist for over 5 years](https://www.linkedin.com/in/iza-rokita/) in a fast-paced tech industry,** to identify the most important data and change them into valuable business insights.

This is part 2 of the series, stay tuned for more! In [part 1](https://www.kaggle.com/code/izzierider/x-education-case1) I've focused on getting myself around **Pandas, NumPy, Jupyter Notebook, and Tableau**, to perform a simple analysis of **how many leads come from which country.** BTW - I would change so much about it right now (why pie chart, instead of geo chart, what was I thinking?!), but I guess this shows my progress so I'm leaving it as a reminder, that everyone has to start somewhere.

Feel free to comment and add your feedback - I'm just starting my career as a Business Intelligence / Data Visualization / Data Analyst and would love to receive some feedback!

[This is the dataset I used for my analysis.](https://www.kaggle.com/datasets/amritachatterjee09/lead-scoring-dataset)

*Enjoy!*

In [168]:
#importing the Python libraries I will need for my EDA process
import numpy as np
import pandas as pd

# Step 1 - Preparing the dataset for the analysis

The goal of this EDA is to **identify the qualities of the personas interested in the X-education software product.** I will look for some patterns, to determine which kinds of people have the highest possibility of converting, and which ones do not. This will be very valuable for designing a Lead Scoring process, which I want to tackle in the nearest future.

In the beginning, I'm going to familiarize myself better with the dataset, to formulate additional questions I will try to find answers to based on the data.

Now I should perform basic data exploration, checking the dataset shape, duplicates, nullvalues, datatypes and more. But I already did that [in the previous X-Education project so you can check it here under STEP 1 section.](https://www.kaggle.com/code/izzierider/x-education-case1)

I also identified the columns I would like to use for my EDA and created the data frame containing only them. Just a reminder, I've chosen columns as below:

**Lead Number** - to identify each lead, *int64*

**Lead Origin** - to explore, is there any connection between the lead sign-up location and the conversion rate, *object*

**Converted** - very important column, showing, did the lead eventually convert or not, *int64*

**Total Visits** - the column import for measuring the online engagement of the lead, *float64*

**Total Time Spent** - the column important for measuring the online engagement of the lead, *int64*

**Page Views Per Visit** - the column important for measuring the online engagement of the lead, *float64*

**Country** - important for persona analysis, *object*

**Specialization** - important for persona analysis, *object*

**Occupation** - important for persona analysis, *object*

**Lead Quality** - important for determining the quality of the contact, *object*


In [169]:
#importing the dataset csv file as a pandas dataframe
df = pd.read_csv('CASE2_df.csv') 

In [170]:
# checking how the first 10 rows look like
df.head(10)

Lead Number              Lead Origin  Converted  TotalVisits   
0       660737                      API          0          0.0  \
1       660728                      API          0          5.0   
2       660727  Landing Page Submission          1          2.0   
3       660719  Landing Page Submission          0          1.0   
4       660681  Landing Page Submission          1          2.0   
5       660680                      API          0          0.0   
6       660673  Landing Page Submission          1          2.0   
7       660664                      API          0          0.0   
8       660624  Landing Page Submission          0          2.0   
9       660616                      API          0          4.0   

   Total Time Spent on Website  Page Views Per Visit Country   
0                            0                   0.0     NaN  \
1                          674                   2.5   India   
2                         1532                   2.0   India   
3                          305                   1.0   India   
4                         1428                   1.0   India   
5                            0                   0.0     NaN   
6                         1640                   2.0   India   
7                            0                   0.0     NaN   
8                           71                   2.0   India   
9                           58                   4.0   India   

            Specialization What is your current occupation      Lead Quality  
0                   Select                      Unemployed  Low in Relevance  
1                   Select                      Unemployed               NaN  
2  Business Administration                         Student          Might be  
3    Media and Advertising                      Unemployed          Not Sure  
4                   Select                      Unemployed          Might be  
5                      NaN                             NaN               NaN  
6  Supply Chain Management                      Unemployed  Low in Relevance  
7                      NaN                             NaN               NaN  
8   IT Projects Management                             NaN               NaN  
9       Finance Management                             NaN               NaN

In [171]:
# checking the dataframe size - 9240 rows and 10 columns
df.shape

(9240, 10)

In [172]:
df.duplicated()

0       False
1       False
2       False
3       False
4       False
        ...  
9235    False
9236    False
9237    False
9238    False
9239    False
Length: 9240, dtype: bool

In [173]:
# identifying datatypes
df.dtypes

Lead Number                          int64
Lead Origin                         object
Converted                            int64
TotalVisits                        float64
Total Time Spent on Website          int64
Page Views Per Visit               float64
Country                             object
Specialization                      object
What is your current occupation     object
Lead Quality                        object
dtype: object

In [174]:
# checking for null values
df.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9240 entries, 0 to 9239
Data columns (total 10 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   Lead Number                      9240 non-null   int64  
 1   Lead Origin                      9240 non-null   object 
 2   Converted                        9240 non-null   int64  
 3   TotalVisits                      9103 non-null   float64
 4   Total Time Spent on Website      9240 non-null   int64  
 5   Page Views Per Visit             9103 non-null   float64
 6   Country                          6779 non-null   object 
 7   Specialization                   7802 non-null   object 
 8   What is your current occupation  6550 non-null   object 
 9   Lead Quality                     4473 non-null   object 
dtypes: float64(2), int64(3), object(5)
memory usage: 722.0+ KB


We can see, that there are some values missing in columns:
*     TotalVisits 
*     Page Views Per Visit
*     Country
*     Specialization
*     What is your current occupation
*     Lead Quality

For this specific EDA it is worth to notice, how many nulls are in columns covering:

    Specialization (1438 missing values)
    
    Current occupation (2690 missing values)
    
    Lead Quality (4767 missing values)

Now it's the time to select specific columns from the dataframe that I will need to get some valuable business insights.

In [175]:
# checking for unique values for the Specialization columns
df['Specialization'].unique()

array(['Select', 'Business Administration', 'Media and Advertising', nan,
       'Supply Chain Management', 'IT Projects Management',
       'Finance Management', 'Travel and Tourism',
       'Human Resource Management', 'Marketing Management',
       'Banking, Investment And Insurance', 'International Business',
       'E-COMMERCE', 'Operations Management', 'Retail Management',
       'Services Excellence', 'Hospitality Management',
       'Rural and Agribusiness', 'Healthcare Management', 'E-Business'],
      dtype=object)

In [176]:
# checking for unique values for the What is your current occupation columns
df['What is your current occupation'].unique()

array(['Unemployed', 'Student', nan, 'Working Professional',
       'Businessman', 'Other', 'Housewife'], dtype=object)

In [177]:
# checking for unique values for the Lead Quality columns
df['Lead Quality'].unique()

array(['Low in Relevance', nan, 'Might be', 'Not Sure', 'Worst',
       'High in Relevance'], dtype=object)

# Step 2 - Creating a new dataframe with required columns

**Lead number** - serving as an ID key, *float64 dtype*
    
**Specialization** - serving to identify the lead quality compared to persona industry, *object dtype*
    
**What is your current occupation** - important to identify the lead quality compared to persona occupation,  *object dtype*
    
**Lead Quality** - based on the data set dictionary, the indicator on how possible it is for lead to converted based on data and the intuition of a person assigned to a lead *object dtype* 
    
**Converted** - important to verify, if the lead quality segmentation is accurate *int 64 dtype*

In [178]:
# creating a custom dataframe with chosen columns
chosen_columns = ['Lead Number', 'Specialization', 'What is your current occupation', 'Lead Quality', 'Converted']
leads_quality_df = df[chosen_columns]

In [179]:
# verifying if it worked
leads_quality_df.head(10)

Lead Number           Specialization What is your current occupation   
0       660737                   Select                      Unemployed  \
1       660728                   Select                      Unemployed   
2       660727  Business Administration                         Student   
3       660719    Media and Advertising                      Unemployed   
4       660681                   Select                      Unemployed   
5       660680                      NaN                             NaN   
6       660673  Supply Chain Management                      Unemployed   
7       660664                      NaN                             NaN   
8       660624   IT Projects Management                             NaN   
9       660616       Finance Management                             NaN   

       Lead Quality  Converted  
0  Low in Relevance          0  
1               NaN          0  
2          Might be          1  
3          Not Sure          0  
4          Might be          1  
5               NaN          0  
6  Low in Relevance          1  
7               NaN          0  
8               NaN          0  
9               NaN          0

In [180]:
leads_quality_df.shape

(9240, 5)

Just for curiosity, I will drop now all rows with null values to see, how much data I will loose.

In [181]:
leads_quality_df.dropna()

Lead Number           Specialization What is your current occupation   
0          660737                   Select                      Unemployed  \
2          660727  Business Administration                         Student   
3          660719    Media and Advertising                      Unemployed   
4          660681                   Select                      Unemployed   
6          660673  Supply Chain Management                      Unemployed   
...           ...                      ...                             ...   
9234       579615  Business Administration                      Unemployed   
9235       579564   IT Projects Management                      Unemployed   
9236       579546    Media and Advertising                      Unemployed   
9237       579545  Business Administration                      Unemployed   
9239       579533  Supply Chain Management                      Unemployed   

           Lead Quality  Converted  
0      Low in Relevance          0  
2              Might be          1  
3              Not Sure          0  
4              Might be          1  
6      Low in Relevance          1  
...                 ...        ...  
9234           Might be          1  
9235  High in Relevance          1  
9236           Might be          0  
9237           Not Sure          0  
9239           Might be          1  

[4473 rows x 5 columns]

As I see, I've lost more than half of my data which is a lot. 
In this case I will replace some NaNs with custom data to make my analysis more relevant. I've learnt that from the  ["Prepare Data for Exploration"](https://www.coursera.org/account/accomplishments/certificate/DKVD54E3DANT) course I finished, provided by Google.

In [182]:
# Now I will check each column for NaN values, and replace them with str value "missing data"
leads_quality_df['Specialization'].unique()

array(['Select', 'Business Administration', 'Media and Advertising', nan,
       'Supply Chain Management', 'IT Projects Management',
       'Finance Management', 'Travel and Tourism',
       'Human Resource Management', 'Marketing Management',
       'Banking, Investment And Insurance', 'International Business',
       'E-COMMERCE', 'Operations Management', 'Retail Management',
       'Services Excellence', 'Hospitality Management',
       'Rural and Agribusiness', 'Healthcare Management', 'E-Business'],
      dtype=object)

In [183]:
# We will count nulls for the Specialization column
specialization_nulls = leads_quality_df['Specialization'].isnull().sum()
print(specialization_nulls)

1438


That is a lot of null values! One of the possible reasons why it happened, is that there is no category "Other" for leads from other industries. Also there is a category 'Select" which doesn't tell us anything. So we can savely replace the value in this column, where the selected Specialization is "Select" and nan values with custom value "Specialization unknown" (str dtype).

In [184]:
column = 'Specialization' 
old_value = 'Select'  
new_value = 'Specialization unknown' 

leads_quality_df[column] = leads_quality_df[column].replace(old_value, new_value)
leads_quality_df[column] = leads_quality_df[column].fillna(new_value)

leads_quality_df_new = leads_quality_df

print(leads_quality_df_new)

      Lead Number             Specialization What is your current occupation   
0          660737     Specialization unknown                      Unemployed  \
1          660728     Specialization unknown                      Unemployed   
2          660727    Business Administration                         Student   
3          660719      Media and Advertising                      Unemployed   
4          660681     Specialization unknown                      Unemployed   
...           ...                        ...                             ...   
9235       579564     IT Projects Management                      Unemployed   
9236       579546      Media and Advertising                      Unemployed   
9237       579545    Business Administration                      Unemployed   
9238       579538  Human Resource Management                             NaN   
9239       579533    Supply Chain Management                      Unemployed   

           Lead Quality  Converted  
0 

/var/folders/46/b24dskq57mzgcckg0dsrr6jr0000gp/T/ipykernel_94449/3534687059.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  leads_quality_df[column] = leads_quality_df[column].replace(old_value, new_value)
/var/folders/46/b24dskq57mzgcckg0dsrr6jr0000gp/T/ipykernel_94449/3534687059.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  leads_quality_df[column] = leads_quality_df[column].fillna(new_value)


In [185]:
# now it's time to check number of nulls after our Data Cleaning
specialization_nulls = leads_quality_df_new['Specialization'].isnull().sum()
print(specialization_nulls)

0


As a result, we got rid of nan values inside the specialization column, now let's see how it influenced the general data frame missing values count.

In [186]:
leads_quality_df_new.dropna()

Lead Number           Specialization What is your current occupation   
0          660737   Specialization unknown                      Unemployed  \
2          660727  Business Administration                         Student   
3          660719    Media and Advertising                      Unemployed   
4          660681   Specialization unknown                      Unemployed   
6          660673  Supply Chain Management                      Unemployed   
...           ...                      ...                             ...   
9234       579615  Business Administration                      Unemployed   
9235       579564   IT Projects Management                      Unemployed   
9236       579546    Media and Advertising                      Unemployed   
9237       579545  Business Administration                      Unemployed   
9239       579533  Supply Chain Management                      Unemployed   

           Lead Quality  Converted  
0      Low in Relevance          0  
2              Might be          1  
3              Not Sure          0  
4              Might be          1  
6      Low in Relevance          1  
...                 ...        ...  
9234           Might be          1  
9235  High in Relevance          1  
9236           Might be          0  
9237           Not Sure          0  
9239           Might be          1  

[4473 rows x 5 columns]

In [187]:
leads_quality_df_new['Specialization'].unique()

array(['Specialization unknown', 'Business Administration',
       'Media and Advertising', 'Supply Chain Management',
       'IT Projects Management', 'Finance Management',
       'Travel and Tourism', 'Human Resource Management',
       'Marketing Management', 'Banking, Investment And Insurance',
       'International Business', 'E-COMMERCE', 'Operations Management',
       'Retail Management', 'Services Excellence',
       'Hospitality Management', 'Rural and Agribusiness',
       'Healthcare Management', 'E-Business'], dtype=object)

The number of data I would loose by dropping all nulls is still really high, but the replacements worked, there are no values like "Select" or nans. So we need to continue with the cleaning! Let's move to the "What is your current occupation" column.

In [188]:
occupation_nulls = leads_quality_df_new['What is your current occupation'].isnull().sum()
print(occupation_nulls)

2690


Even more nulls in this category! Let's see if we can replace them somehow.

In [189]:
leads_quality_df_new['What is your current occupation'].unique()

array(['Unemployed', 'Student', nan, 'Working Professional',
       'Businessman', 'Other', 'Housewife'], dtype=object)

In this case, we have a category "Other" so that's a good sign. Yet over 1/4 of the leads didn't fill that field at all! It's worth to investigate it further - especially if they converted. But for now, we will replace all nulls with the "Undisclosed" value.

In [190]:
column2 = 'What is your current occupation' 
new_value2 = 'Undisclosed' 

leads_quality_df_new[column2] = leads_quality_df[column2].fillna(new_value2)

leads_quality_df = leads_quality_df_new

print(leads_quality_df)

      Lead Number             Specialization What is your current occupation   
0          660737     Specialization unknown                      Unemployed  \
1          660728     Specialization unknown                      Unemployed   
2          660727    Business Administration                         Student   
3          660719      Media and Advertising                      Unemployed   
4          660681     Specialization unknown                      Unemployed   
...           ...                        ...                             ...   
9235       579564     IT Projects Management                      Unemployed   
9236       579546      Media and Advertising                      Unemployed   
9237       579545    Business Administration                      Unemployed   
9238       579538  Human Resource Management                     Undisclosed   
9239       579533    Supply Chain Management                      Unemployed   

           Lead Quality  Converted  
0 

/var/folders/46/b24dskq57mzgcckg0dsrr6jr0000gp/T/ipykernel_94449/1278959909.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  leads_quality_df_new[column2] = leads_quality_df[column2].fillna(new_value2)


In [191]:
# now let's check if I got rid of the nulls in the What is your current occupation section - yes I did!
occupation_nulls = leads_quality_df['What is your current occupation'].isnull().sum()
print(occupation_nulls)

0


In [192]:
# double check - yep, there is new value "Undisclosed" instead of nan as we wanted
leads_quality_df['What is your current occupation'].unique()

array(['Unemployed', 'Student', 'Undisclosed', 'Working Professional',
       'Businessman', 'Other', 'Housewife'], dtype=object)

In [193]:
leads_quality_df.dropna()

Lead Number           Specialization What is your current occupation   
0          660737   Specialization unknown                      Unemployed  \
2          660727  Business Administration                         Student   
3          660719    Media and Advertising                      Unemployed   
4          660681   Specialization unknown                      Unemployed   
6          660673  Supply Chain Management                      Unemployed   
...           ...                      ...                             ...   
9234       579615  Business Administration                      Unemployed   
9235       579564   IT Projects Management                      Unemployed   
9236       579546    Media and Advertising                      Unemployed   
9237       579545  Business Administration                      Unemployed   
9239       579533  Supply Chain Management                      Unemployed   

           Lead Quality  Converted  
0      Low in Relevance          0  
2              Might be          1  
3              Not Sure          0  
4              Might be          1  
6      Low in Relevance          1  
...                 ...        ...  
9234           Might be          1  
9235  High in Relevance          1  
9236           Might be          0  
9237           Not Sure          0  
9239           Might be          1  

[4473 rows x 5 columns]

Still, there are lots of nulls remaining - so moving on with cleaning! Now it's time for the Lead Quality column.

In [194]:
lead_quality_nulls = leads_quality_df['Lead Quality'].isnull().sum()
print(lead_quality_nulls)

4767


Ok, we found the reason why we lost so much data if we drop all nulls from the data frame. Around half of the data don't have the column "Lead quality" filled. Let's see what we can do about it.

In [195]:
leads_quality_df['Lead Quality'].unique()

array(['Low in Relevance', nan, 'Might be', 'Not Sure', 'Worst',
       'High in Relevance'], dtype=object)

### *Authour insights*

The categories inside the "Lead Quality" column are not set properly. For example, I don't know, what is the clear difference between "Might be" and "Not sure". And "worst" sounds a bit aggressive :) I would personally recommend building a lead scoring strategy and setting clear factors that influence categorizing the lead quality one way or another. 

**For example: determining around 10 indicators and giving points based on the info we have about the lead, how they align with our general business strategy, and the historical data about the client persona, that brought us the most profit in the past.**

This could ensure the data consistency and make sure, that the answers like "might be" or "not sure" would be eliminated, not mentioning the nans. 

What I have in mind - if the lead is Businessman, we give them 3 points, if they are unemployed - 1. If their specialization is project management - they receive 3 points, if hospitality - 1. 

Then we can determine ranges like 0 - 10 points = low quality, 10 - 20 = mediocre quality, 20 - 30 = high quality, and 30 - 40 = highest quality. Of course, this is just an example, and many factors and data need to be taken into consideration and the whole process must be aligned and understandable for various business stakeholders. 

As a Business Intelligence / Data Analyst - I would gladly take this responsibility upon myself.

But for now - back to the data! As the reality is, that we have lots of nans - we have to replace them with something. Let's use the "Unqualified" term.

In [196]:
column3 = 'Lead Quality' 
new_value3 = 'Unqualified' 

leads_quality_df[column3] = leads_quality_df[column3].fillna(new_value3)

leads_quality_df_new = leads_quality_df

print(leads_quality_df_new)

      Lead Number             Specialization What is your current occupation   
0          660737     Specialization unknown                      Unemployed  \
1          660728     Specialization unknown                      Unemployed   
2          660727    Business Administration                         Student   
3          660719      Media and Advertising                      Unemployed   
4          660681     Specialization unknown                      Unemployed   
...           ...                        ...                             ...   
9235       579564     IT Projects Management                      Unemployed   
9236       579546      Media and Advertising                      Unemployed   
9237       579545    Business Administration                      Unemployed   
9238       579538  Human Resource Management                     Undisclosed   
9239       579533    Supply Chain Management                      Unemployed   

           Lead Quality  Converted  
0 

/var/folders/46/b24dskq57mzgcckg0dsrr6jr0000gp/T/ipykernel_94449/1005285255.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  leads_quality_df[column3] = leads_quality_df[column3].fillna(new_value3)


In [197]:
#let's check if we got rid of nulls in this column
lead_quality_nulls = leads_quality_df_new['Lead Quality'].isnull().sum()
print(lead_quality_nulls)

0


In [198]:
# new column value has arrived - so everything went according to the plan!
leads_quality_df_new['Lead Quality'].unique()

array(['Low in Relevance', 'Unqualified', 'Might be', 'Not Sure', 'Worst',
       'High in Relevance'], dtype=object)

In [199]:
# time to check what is the general NaN situation inside our dataframe!
leads_quality_df_new.dropna()

Lead Number             Specialization What is your current occupation   
0          660737     Specialization unknown                      Unemployed  \
1          660728     Specialization unknown                      Unemployed   
2          660727    Business Administration                         Student   
3          660719      Media and Advertising                      Unemployed   
4          660681     Specialization unknown                      Unemployed   
...           ...                        ...                             ...   
9235       579564     IT Projects Management                      Unemployed   
9236       579546      Media and Advertising                      Unemployed   
9237       579545    Business Administration                      Unemployed   
9238       579538  Human Resource Management                     Undisclosed   
9239       579533    Supply Chain Management                      Unemployed   

           Lead Quality  Converted  
0      Low in Relevance          0  
1           Unqualified          0  
2              Might be          1  
3              Not Sure          0  
4              Might be          1  
...                 ...        ...  
9235  High in Relevance          1  
9236           Might be          0  
9237           Not Sure          0  
9238        Unqualified          1  
9239           Might be          1  

[9240 rows x 5 columns]

In [200]:
leads_quality_df.isnull().sum()

Lead Number                        0
Specialization                     0
What is your current occupation    0
Lead Quality                       0
Converted                          0
dtype: int64

Mission accomplished! Our dataframe is cleared out from the null values without loosing any data rows ;)

But just to be sure, let's check every important data once again!

In [201]:
#the shape is right!
leads_quality_df.shape

(9240, 5)

In [202]:
leads_quality_df['Specialization'].unique()

array(['Specialization unknown', 'Business Administration',
       'Media and Advertising', 'Supply Chain Management',
       'IT Projects Management', 'Finance Management',
       'Travel and Tourism', 'Human Resource Management',
       'Marketing Management', 'Banking, Investment And Insurance',
       'International Business', 'E-COMMERCE', 'Operations Management',
       'Retail Management', 'Services Excellence',
       'Hospitality Management', 'Rural and Agribusiness',
       'Healthcare Management', 'E-Business'], dtype=object)

In [203]:
leads_quality_df['What is your current occupation'].unique()

array(['Unemployed', 'Student', 'Undisclosed', 'Working Professional',
       'Businessman', 'Other', 'Housewife'], dtype=object)

In [204]:
leads_quality_df['Lead Quality'].unique()

array(['Low in Relevance', 'Unqualified', 'Might be', 'Not Sure', 'Worst',
       'High in Relevance'], dtype=object)

In [205]:
leads_quality_df['Converted'].unique()

array([0, 1])

In [206]:
leads_quality_df.dtypes

Lead Number                         int64
Specialization                     object
What is your current occupation    object
Lead Quality                       object
Converted                           int64
dtype: object

It looks like everything is allright! So save our cleaned dataframe as a csv for later, and now - the Visualizations time ;)

In [207]:
leads_quality_df.to_csv('x_education_case2_cleaned_data.csv', index=False) 

# Step 3 - Creating two separate datasets for easier analysis

I like order so I want to create two separate datasets for my further analysis. One containing leads that converted and second one, where I will store those who did not.

In [208]:
df = leads_quality_df

In [209]:
# Checking how many rows each dataset will have
overall_number_converted_leads = df['Converted'].value_counts()[1]
overall_number_unconverted_leads = df['Converted'].value_counts()[0]
print("Number of converted leads in dataset:", overall_number_converted_leads)
print("Number of leads that didn't convert:", overall_number_unconverted_leads)

Number of converted leads in dataset: 3561
Number of leads that didn't convert: 5679


**First conclusion** 
Among the whole dataset, **37,80 % of leads converted**, and **62,20 % didn't** *(rounding up to 2 digits after coma)*. The size of each dataset is big enough, to be analyzed. For example - if the overall number of converted leads would be 500 and those who did not 9000, I would focus on analyzing only the leads that did not convert, as the size of the converted leads dataset would be to small.

In [210]:
# We're creating a data frame only with leads that converted
df_converted = df[df['Converted'] == 1]

In [211]:
# It worked - number of rows is the same as the overall number of converted leads we counted above
df_converted.shape

(3561, 5)

In [212]:
# Let's print the new dataframe, to be absolutely sure - yep, it worked
df_converted.head(20)

Lead Number                     Specialization   
2        660727            Business Administration  \
4        660681             Specialization unknown   
6        660673            Supply Chain Management   
10       660608                 Travel and Tourism   
11       660570          Human Resource Management   
12       660562               Marketing Management   
15       660547             Specialization unknown   
18       660522               Marketing Management   
22       660471  Banking, Investment And Insurance   
24       660458               Marketing Management   
25       660447                 Travel and Tourism   
26       660432                 Finance Management   
27       660424              Media and Advertising   
29       660410             International Business   
30       660402          Human Resource Management   
34       660333                         E-COMMERCE   
35       660321          Human Resource Management   
37       660267                 Finance Management   
39       660236              Operations Management   
50       660087             Specialization unknown   

   What is your current occupation       Lead Quality  Converted  
2                          Student           Might be          1  
4                       Unemployed           Might be          1  
6                       Unemployed   Low in Relevance          1  
10                      Unemployed           Might be          1  
11                      Unemployed        Unqualified          1  
12                      Unemployed           Might be          1  
15                      Unemployed           Might be          1  
18                      Unemployed        Unqualified          1  
22                      Unemployed   Low in Relevance          1  
24                      Unemployed  High in Relevance          1  
25                      Unemployed           Not Sure          1  
26                      Unemployed   Low in Relevance          1  
27                      Unemployed  High in Relevance          1  
29                      Unemployed  High in Relevance          1  
30                      Unemployed           Might be          1  
34                      Unemployed           Might be          1  
35                      Unemployed  High in Relevance          1  
37            Working Professional   Low in Relevance          1  
39                      Unemployed  High in Relevance          1  
50                      Unemployed   Low in Relevance          1

In [213]:
#Making absolutely sure, only converted leads are on the list
df_converted['Converted'].unique()

array([1])

Great, so now let's create a similar data frame, but with leads that did not convert.

In [214]:
# We're creating a data frame only with leads that did not convert
df_not_converted = df[df['Converted'] == 0]

In [215]:
# It worked - number of rows is the same as the overall number of not converted leads we counted earlier
df_not_converted.shape

(5679, 5)

In [216]:
df_not_converted['Converted'].unique()

array([0])

As a result, we have 2 new dataframes:

df_converted = it stores all leads that converted

df_not_converted = it stores all leads that did not convert

Let's save them as separate CSVs as well.

In [217]:
#We're creating a dataset with only leads that converted in csv file
df_converted.to_csv('converted_leads_target_persona_04_07_2023.csv', index=False) 

In [218]:
#We're creating a dataset with only leads that did not convert in csv file
df_not_converted.to_csv('not_converted_leads_target_persona_04_07_2023.csv', index=False) 

Awesome! Let's take those two datasets to MySQL now and blend them with the country data from the original dataset for deeper understanding of the leads characteristics.



# Step 4 - Data Blending using SQL (MySQL)

Let's alter column names in merged tables, to make it easier to blend. I've uploaded my csv files to a MySQL database, so the code won't obviously work here (it worked in MySQL, I promise!), but I'm pasting it anyway to show my thought process. It's in Markdown.

Starting out with changing the column names, to make blending much easier.

ALTER TABLE `Trening_SQL`.`converted_leads` 
CHANGE COLUMN `Lead Number` `Lead_Number` INT NULL DEFAULT NULL ;

ALTER TABLE `Trening_SQL`.`converted_leads` 
CHANGE COLUMN `What is your current occupation` `Current_occupation` INT NULL DEFAULT NULL ;

ALTER TABLE `Trening_SQL`.`converted_leads` 
CHANGE COLUMN `Lead Quality` `Lead_Quality` INT NULL DEFAULT NULL ;

ALTER TABLE `Trening_SQL`.`lead_scoring` 
CHANGE COLUMN `Lead Number` `Lead_Number` INT NULL DEFAULT NULL ;

ALTER TABLE `Trening_SQL`.`not_converted_leads` 
CHANGE COLUMN `Lead Number` `Lead_Number` INT NULL DEFAULT NULL ;

ALTER TABLE `Trening_SQL`.`not_converted_leads` 
CHANGE COLUMN `What is your current occupation` `Current_occupation` TEXT NULL DEFAULT NULL ;

ALTER TABLE `Trening_SQL`.`not_converted_leads` 
CHANGE COLUMN `Lead Quality` `Lead_Quality` TEXT NULL DEFAULT NULL ;

Now, let's use the SQL Left join method, to receive a table where all of the records from the converted leads table will be available, 
and the country data from the original dataset. They will be matched on the primary key "Lead_Number". 

SELECT 
    `Trening_SQL`.`converted_leads`.Lead_Number,
    Specialization,
    Current_occupation,
    Lead_Quality,
    Country
FROM
    `Trening_SQL`.`converted_leads`
        LEFT JOIN
    `Trening_SQL`.`lead_scoring` ON `Trening_SQL`.`converted_leads`.Lead_Number = `Trening_SQL`.`lead_scoring`.Lead_Number;


The number of rows is still 3561, so we did not loose any data in the process. Let's do the same with leads that did not convert.


SELECT 
    `Trening_SQL`.`not_converted_leads`.Lead_Number,
    Specialization,
    Current_occupation,
    Lead_Quality,
    Country
FROM
    `Trening_SQL`.`not_converted_leads`
        LEFT JOIN
    `Trening_SQL`.`lead_scoring` ON `Trening_SQL`.`not_converted_leads`.Lead_Number = `Trening_SQL`.`lead_scoring`.Lead_Number;

Success once again! With 5679 rows available after blending, we have our two tables extended by country info.

Let's load them as dataframes and take care of newly appeared null values.

In [219]:
# Importing blended data into dataframes
converted = pd.read_csv('converted_leads_blended.csv')

In [220]:
not_converted = pd.read_csv('not_converted_leads_blended.csv') 

In [221]:
#checking 20 rows for converted leads data set, is the Country column there
converted.head(20)

Lead_Number                     Specialization    Current_occupation   
0        660727            Business Administration               Student  \
1        660681             Specialization unknown            Unemployed   
2        660673            Supply Chain Management            Unemployed   
3        660608                 Travel and Tourism            Unemployed   
4        660570          Human Resource Management            Unemployed   
5        660562               Marketing Management            Unemployed   
6        660547             Specialization unknown            Unemployed   
7        660522               Marketing Management            Unemployed   
8        660471  Banking, Investment And Insurance            Unemployed   
9        660458               Marketing Management            Unemployed   
10       660447                 Travel and Tourism            Unemployed   
11       660432                 Finance Management            Unemployed   
12       660424              Media and Advertising            Unemployed   
13       660410             International Business            Unemployed   
14       660402          Human Resource Management            Unemployed   
15       660333                         E-COMMERCE            Unemployed   
16       660321          Human Resource Management            Unemployed   
17       660267                 Finance Management  Working Professional   
18       660236              Operations Management            Unemployed   
19       660087             Specialization unknown            Unemployed   

         Lead_Quality  Country  
0            Might be    India  
1            Might be    India  
2    Low in Relevance    India  
3            Might be    India  
4         Unqualified    India  
5            Might be    India  
6            Might be    India  
7         Unqualified    India  
8    Low in Relevance    India  
9   High in Relevance    India  
10           Not Sure    India  
11   Low in Relevance    India  
12  High in Relevance    India  
13  High in Relevance    India  
14           Might be    India  
15           Might be    India  
16  High in Relevance    India  
17   Low in Relevance    India  
18  High in Relevance    India  
19   Low in Relevance  unknown

In [222]:
# Looking for nulls
converted['Country'].unique()

array(['India', 'unknown', 'United States', 'Bahrain', 'Singapore',
       'United Kingdom', 'United Arab Emirates', 'Qatar', 'Saudi Arabia',
       'France', 'Netherlands', 'Germany', 'Sweden', 'Oman',
       'Asia/Pacific Region', 'Hong Kong', 'Denmark', 'South Africa',
       'Australia', 'Bangladesh'], dtype=object)

In [223]:
#  there are few of them, let's replace them with the "unknown" value and check if it work
column = 'Country' 
new_value = 'unknown' 

converted[column] = converted[column].fillna(new_value)

converted_new = converted

converted_new['Country'].unique()

array(['India', 'unknown', 'United States', 'Bahrain', 'Singapore',
       'United Kingdom', 'United Arab Emirates', 'Qatar', 'Saudi Arabia',
       'France', 'Netherlands', 'Germany', 'Sweden', 'Oman',
       'Asia/Pacific Region', 'Hong Kong', 'Denmark', 'South Africa',
       'Australia', 'Bangladesh'], dtype=object)

In [224]:
# It worked, so let's do the same for the not converted leads dataset

column = 'Country' 
new_value = 'unknown' 

not_converted[column] = not_converted[column].fillna(new_value)

not_converted_new = not_converted

not_converted_new['Country'].unique()

array(['unknown', 'India', 'Russia', 'Kuwait', 'Oman',
       'United Arab Emirates', 'United States', 'Australia',
       'United Kingdom', 'Ghana', 'Singapore', 'Qatar', 'Saudi Arabia',
       'Belgium', 'France', 'Sri Lanka', 'China', 'Canada', 'Sweden',
       'Nigeria', 'Hong Kong', 'Uganda', 'Germany', 'Kenya', 'Italy',
       'Bahrain', 'South Africa', 'Tanzania', 'Malaysia', 'Netherlands',
       'Liberia', 'Switzerland', 'Philippines', 'Bangladesh', 'Vietnam',
       'Indonesia', 'Asia/Pacific Region'], dtype=object)

In [225]:
# saving dataframes as csv files (replacing the old ones)

converted_new.to_csv('converted_leads_blended.csv', index=False)
not_converted_new.to_csv('not_converted_leads_blended.csv', index=False) 

# Step 5 - Data Visualization with Tableau

The easiest way to get the specific insights from a Cleaned and Initially Transformed data is to use a Data Visualization tool. Recently I've finished this amazing course titled [Go Beyond the Numbers: Translate Data into Insights from Google](https://www.coursera.org/account/accomplishments/certificate/5KQ2FUXV5QYL) where I could get myself familiar with [Tableau Public Data Visualization tool](https://public.tableau.com/). 

I liked it so much, that I've enrolled in a more complex course on that matter on Udemy - [Tableau 2022 A-Z: Hands-On Tableau Training for Data Science](https://www.udemy.com/course/tableau10/).

### **Based on the tips from those courses, I've prepared Data Visualizations available on my [Tableau Public profile here](https://public.tableau.com/app/profile/izabela.rokita).**


# Conclusion

The ultimate goal of any Exploratory Data Analysis process, **should be getting valuable Business Insights.** Asking the right questions, deciding which data we need (and which we don't), and presenting our discoveries in an appealing to non-tech stakeholders way. 

Based on my whole analysis I found insights below:

### #1 The Lead scoring process needs a complete reconstruction

Based on the data, the sales team have got trouble with qualifying the lead quality accordingly. Among the converted leads, the most common categories are "Might be" and "Unqualified" - not "High in Relevance". On the opposite - among not converted leads leading categories are "Unqualified" and "Not sure", not "Worst". Overall, the number of "unqualified" leads is also concerning. To improve the conversion rate and allocate the sales team efforts in the right places, it would be beneficial to redesign the whole lead scoring strategy,

### #2 Visible difficulties to acquire and convert leads from other countries than India

As the Data Visualizations show, in terms of geography, India dominates in terms of leads converted. Despite the fact, of being a very big market, it would beneficial to diversify the target geography more. There are a few leads from countries like the USA, Australia, or Saudi Arabia, but not enough, to make a real difference. My recommendation is, to invest in market research in different regions, to identify where the X Education product could fulfill some niche. Then, work on a targeted offer, to satisfy clients from specific regions' needs.

### #3 Leads top 3 specializations are the same both for converted and not converted leads

In both cases, Finance Management, Marketing Management, and Human Resources Management Specialization types are chosen most often. A good idea would be, to prepare some personalized offers for those domains, and focus 80 % of the marketing/sales forces on them (Pareto rule).

### #4 X Education product is most popular among people who are currently Unemployed

Adjusting tone of voice and promotional educational materials to people looking for a job, could improve brand awareness and help with spreading the word about the product. Yet don't forget about the Working Professionals, as they can have higher Lead Value, despite that there are fewer of them.

**If you got to this point - thank you for your time! [Here is my LinkedIn profile](https://www.linkedin.com/in/iza-rokita/), feel free to connect in any matter relevant. Have an insightful day!**